### PR TITLE
[SPARK-47250][SS] Add additional validations and NERF changes for RocksDB state provider and use of column families

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add additional validations and NERF changes for RocksDB state provider and use of col families

### Why are the changes needed?
Improve error handling and migrating errors to NERF. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added new unit tests
StateStoreSuite
```
===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.execution.streaming.state.StateStoreSuite, threads: shuffle-boss-36-1 (daemon=true), ForkJoinPool.commonPool-worker-1 (daemon=true), ForkJoinPool.commonPool-worker-3 (daemon=true), rpc-boss-33-1 (daemon=true), ForkJoinPool.commonPool-worker-2 (daemon=true) =====
[info] Run completed in 2 minutes, 57 seconds.
[info] Total number of tests run: 151
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 151, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

RocksDBSuite
```
[info] Run completed in 4 minutes, 54 seconds.
[info] Total number of tests run: 188
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 188, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
@sahnib @HeartSaVioR - PTAL, thx !

@HeartSaVioR - let me know if you are ok with the proposed dir layout changes and also if you prefer them in a separate PR. Thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 1]
```diff
@@ -246,25 +246,35 @@ class RocksDB(
...
   }
 
-  private def verifyColFamilyExists(colFamilyName: String): Unit = {
-    if (useColumnFamilies && !checkColFamilyExists(colFamilyName)) {
-      throw new RuntimeException(s"Column family with name=$colFamilyName does not exist")
+  private def verifyColFamilyExists(operationName: String, colFamilyName: String): Unit = {
+    if (colFamilyName != StateStore.DEFAULT_COL_FAMILY_NAME) {
+      if (!useColumnFamilies) {
+        throw StateStoreErrors.unsupportedOperationException(operationName,
+          "RocksDBStateStoreProvider and multiple column families disabled")
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:253]: can be reworded as - `multiple column families disabled in RocksDBStateStoreProvider`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 2]
```diff
@@ -246,25 +246,35 @@ class RocksDB(
...
   /**
    * Create RocksDB column family, if not created already
    */
   def createColFamilyIfAbsent(colFamilyName: String): Unit = {
-    if (colFamilyName == StateStore.DEFAULT_COL_FAMILY_NAME) {
-      throw new SparkUnsupportedOperationException(
-        errorClass = "_LEGACY_ERROR_TEMP_3197",
-        messageParameters = Map("colFamilyName" -> colFamilyName).toMap)
+    // Remove leading and trailing whitespaces
+    val cfName = colFamilyName.trim
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:268]: This trimming also needs to happen when the cx issues a put/get etc against the column family. I dont think thats happening today. In that case, would this lead to inconsistencies?

**Review by sahnib — COMMENTED** [review]
Thanks for making these changes. Left a few comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:268]: Hmm - user facing APIs shouldn't actually call state store operations directly by passing the name. The assumption is we should never have col families created with leading/trailing spaces. Do you think we should add an assert to verify the same and throw an exception ?

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala]: Modified to check and throw an exception if some invariants are not met for both the col family operations case and the creation/deletion case

**Review by sahnib — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Is this good to go? @HeartSaVioR @rangadi

**Comment by HeartSaVioR** [comment]
Will review sooner than later. Maybe by today.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 3]
```diff
@@ -536,6 +536,67 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
 
+  testWithColumnFamilies(s"RocksDB: column family creation with invalid names",
+    TestWithBothChangelogCheckpointingEnabledAndDisabled) { colFamiliesEnabled =>
+    val remoteDir = Utils.createTempDir().toString
+    new File(remoteDir).delete() // to make sure that the directory gets created
+
+    val conf = RocksDBConf().copy()
+    withDB(remoteDir, conf = conf, useColumnFamilies = colFamiliesEnabled) { db =>
+      Seq("default", "", " ", "    ", " default", " default ").foreach { colFamilyName =>
+        val ex = intercept[Exception] {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:547]: Verifying exception belongs to error class framework needs be done with checkError.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 4]
```diff
@@ -536,6 +536,67 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
+        }
+        ex.getCause.isInstanceOf[UnsupportedOperationException]
+      }
+    }
+  }
+
+  private def verifyStoreOperationUnsupported(
+      operationName: String)
+      (testFn: => Unit): Unit = {
+    val ex = intercept[UnsupportedOperationException] {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:558]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 5]
```diff
@@ -582,7 +636,7 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
         assert(iterator(db, colFamily2).isEmpty)
       }
       assert(ex.isInstanceOf[RuntimeException])
-      assert(ex.getMessage.contains("does not exist"))
+      assert(ex.getMessage.contains("missing column family"))
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:639]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala [hunk 6]
```diff
@@ -134,6 +134,46 @@ class StateStoreSuite extends StateStoreSuiteBase[HDFSBackedStateStoreProvider]
     }
   }
 
+  private def verifyStoreOperationUnsupported()(testFn: => Unit): Unit = {
+    val ex = intercept[UnsupportedOperationException] {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:138]: ditto

**Review by HeartSaVioR — COMMENTED** [review]
Left minors but probably need to be fixed to respect the policy of error class framework.

Also, I'd like to see refactoring (moving package) as a separate PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala (hunk 3 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala (hunk 4 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala (hunk 5 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala (hunk 6 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala]: Done

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**